Make ignored users in User Operator configurable and extend them to qoutas and credentials (#164)
strimzi/proposals: 106-make-ignored-users-in-UO-configurable.md

And unlike ACL rules created for regular users, the User Operator would just ignore these rules.

To further increase the risk, this mechanism is documented only briefly in a single paragraph in our documentation.
That makes it easy to miss, and most Strimzi users probably don't know about this _feature_.
And while the User Operator logs an INFO level message when it is ignoring any ACL rules, the log message can be easily missed.

This proposal suggests making the list of ignored users configurable with an empty default value (i.e., no users will be ignored by default).
For most users, this should have zero impact.
And users who need the `ANONYMOUS` or `*` users to be ignored can configure them explicitly in the configuration.
The explicit configuration will ensure that they are aware that the ACL rules for these users would be ignored even without carefully reading every single documentation page, because they had to configure it.

In addition to the main motivation - security - this proposal should deliver some additional benefits.

> It proposes extending the ignore list to the whole User Operator and using it even for quotas and SCRAM-SHA-512 credentials.

Reviewer: Shall we say why we're extending here? Ensure consistency or managing users with standalone operator (as mentioned in rejected alternatives)

It also proposes using a regular expression pattern instead of a list of users to allow more flexibility in how the ignored users are defined.
This would give our users more flexibility when using the User Operator in standalone mode.
They would be able to use the User Operator with existing Kafka clusters that already have some other users defined by configuring the ignore pattern to ignore the internal users and not interfere with them while managing other users through the User Operator and `KafkaUser` resources.

## Proposal

A new configuration option will be introduced in the User Operator.
It will be named `STRIMZI_IGNORED_USERS_PATTERN`.

> Its default value will be `null`.

Reviewer: should we mention behavior when not set?

The reconciliation of such a user would proceed normally, and the User Operator will manage the ACLs, quotas, and credentials for such a user.
This is because it will not be affected by the user not being returned by the `getAllUsers` method (we will know about this user from the custom resources).

However, it might be unpredictable what happens when such a `KafkaUser` resource is deleted.
When the User Operator catches the deletion event and reconciles it, it will delete the user completely.
But if the user is deleted while the User Operator is not running or is otherwise busy (full event queue, ongoing reconciliation for given user, etc.), the deletion of the user will be just ignored, and all ACLs, quotas, and credentials will remain in place.

The situation when the user has `KafkaUser` resources with names matching the ignored users pattern is considered a _user configuration error_, and the unpredictable results are considered acceptable behavior.

## Performance impact

For most users, no performance impact is expected as they would not be using this feature at all.

> For users who will use this feature in the _legacy_ mode to ignore the `ANONYMOUS` and `*` usernames, no major impact is expected either.

Reviewer: ```suggestion
For users who enable this feature to ignore `ANONYMOUS` and `*` usernames (as with the legacy approach), no major impact is expected either.
```

For users who decide to use this feature with standalone User Operator and existing Kafka cluster not managed by Strimzi, the performance impact will depend on the complexity of the ignore pattern and the number of ACLs / quotas / credentials that will be ignored.
But in this case, this can be considered as a corresponding cost for this feature.
And given we in general don't expect the User Operator to manage thousands of users, it should not make this feature unusable.

## Documentation changes

When implementing the proposal, the release notes will contain a warning about this change including the way to _reenable_ the original feature by configuring the `STRIMZI_IGNORED_USERS_PATTERN` option.
It is also recommended to mention this change in the _What's new_ video or blog post we usually do for new releases.
This should help ensure that users are aware of this change.

> The regular documentation will be updated to describe the new feature as well.

Reviewer: should we mention migration path to retain old behavior
